src/api/mc_worker_api.erl

```diff
@@ -177,44 +299,134 @@ count(Connection, Query) ->
 
-%% @doc Create index on collection according to given spec.
+%% @doc Create index on collection according to given spec. This function does
+%% not work if you have configured the driver to use the new version of the
+%% protocol with application:set_env(mongodb, use_legacy_protocol, false). In
+%% that case you can call the createIndexes
+%% (https://www.mongodb.com/docs/manual/reference/command/createIndexes/#mongodb-dbcommand-dbcmd.createIndexes)
+%% command using the `mc_worker_api:command/2` function instead. 
+%%
 %%      The key specification is a bson documents with the following fields:
 %%      IndexSpec      :: bson document, for e.g. {field, 1, other, -1, location, 2d}, <strong>required</strong>
 -spec ensure_index(pid(), colldb(), bson:document()) -> ok | {error, any()}.
 ensure_index(Connection, Coll, IndexSpec) ->
-  mc_connection_man:request_worker(Connection, #ensure_index{collection = Coll, index_spec = IndexSpec}).
+    case mc_utils:use_legacy_protocol(Connection) of
+        true ->
+            mc_connection_man:request_worker(Connection,
+                                             #ensure_index{collection = Coll,
+                                                           index_spec = IndexSpec});
+        false -> 
+           erlang:error({error, <<"This function does not work when one have specified application:set_env(mongodb, use_legacy_protocol, false). Call the createIndexes command using mc_worker_api:command/2 instead.">>}) 
+    end.
 
 %% @doc Execute given MongoDB command and return its result.
 -spec command(pid(), mc_worker_api:selector()) -> {boolean(), map()}. % Action
 command(Connection, Query) when is_record(Query, query) ->
-  Doc = mc_connection_man:read_one(Connection, Query),
-  mc_connection_man:process_reply(Doc, Query);
-command(Connection, Command) ->
-  command(Connection,
-    #'query'{
-      collection = <<"$cmd">>,
-      selector = Command
-    }).
+      case mc_utils:use_legacy_protocol(Connection) of
+          true -> 
+              Doc = mc_connection_man:read_one(Connection, Query),
+              mc_connection_man:process_reply(Doc, Query);
+          false ->
+              %% We will convert the legacy command to a modern one.
+              #query{
+                 slaveok = SlaveOk,
+                 selector = Selector} = Query,
+              Fields = bson:fields(Selector),
+              NewSelector =
+              case {lists:keyfind(<<"$readPreference">>, 1, Fields), SlaveOk} of
+                  {{<<"$readPreference">>, _}, _} -> Selector;
+                  {false, true} -> 
+                      bson:document(Fields ++ [{<<"$readPreference">>, #{<<"mode">> => <<"primaryPreferred">>}}]);
+                  {false, false} ->
+                      %% primary is the default mode so we do not need to change anything
+                      Selector
+              end,    
+              command(Connection, NewSelector)
+      end;
+
```

Comment:
```suggestion
```

Reply:
My bad. Good catch! (fixed in commit below)